design/src/rfcs/rfc0023_sdk_credential_cache_type_safety.md

```diff
@@ -0,0 +1,328 @@
+RFC: SDK Credential Cache Type Safety
+=====================================
+
+> Status: RFC
+>
+> Applies to: AWS SDK for Rust
+
+At time of writing (2022-10-11), the SDK's credentials provider can be customized by providing:
+
+1. A profile credentials file to modify the default provider chain
+2. An instance of one of the credentials providers implemented in `aws-config`, such as
+   the `AssumeRoleCredentialsProvider`, `ImdsCredentialsProvider`, and so on.
+3. A custom struct that implements the `ProvideCredentials`
+
+The problem this RFC examines is that when options 2 and 3 above are exercised, the customer
+needs to be aware of credentials caching and put additional effort to ensure caching is set up
+correctly (and that double caching doesn't occur). This is especially difficult to get right
+since some built-in credentials providers (such as `AssumeRoleCredentialsProvider`) already
+have caching, while most others do not and need to be wrapped in `LazyCachingCredentialsProvider`.
+
+The goal of this RFC is to create an API where Rust's type system ensures caching is set up
+correctly, or explicitly opted out of.
+
+`CredentialsCache` and `ConfigLoader::credentials_cache`
+--------------------------------------------------------
+
+A new config method named `credentials_cache()` will be added to `ConfigLoader` and the
+generated service `Config` builders that takes a `CredentialsCache` instance. This `CredentialsCache`
+will be a struct with several functions on it to create and configure the cache.
+
+Client creation will ultimately be responsible for taking this `CredentialsCache` instance
+and wrapping the given (or default) credentials provider.
+
+The `CredentialsCache` would look as follows:
+
+```rust
+enum Inner {
+    Lazy(LazyConfig),
+    // Eager doesn't exist today, so this is purely for illustration
+    Eager(EagerConfig),
+    // Custom may not be implemented right away
+    // Not naming or specifying the custom cache trait for now since its out of scope
+    Custom(Box<dyn SomeCacheTrait>),
+
+    NoCaching,
+}
+pub struct CredentialsCache {
+    inner: Inner,
+}
+
+impl CredentialsCache {
+    // Methods prefixed with `default_` just use the default cache settings
+    pub fn default_lazy() -> Self { /* ... */ }
+    pub fn default_eager() -> Self { /* ... */ }
+
+    // Unprefixed methods return a builder that can take customizations
+    pub fn lazy() -> LazyBuilder { /* ... */ }
+    pub fn eager() -> EagerBuilder { /* ... */ }
+
+    // Later, when custom implementations are supported
+    pub fn custom(cache_impl: Box<dyn SomeCacheTrait>) -> Self { /* ... */ }
+
+    pub(crate) fn create_cache(
```

Comment:
Given this method returns `SharedCredentialsProvider`, maybe we should name this differently? I don't have a name that's spot on, but here is a list of names I'm considering:
- `back_credentials_provider`
- `cache_aware_credentials_provider`
- `enable_cache_for`

Reply:
This method is `pub(crate)`, so the name from the RFC is just a suggestion and doesn't need to match the implementation.

Reply:
Thanks. I missed this comment from Option C, which I assume is still applicable to the selected design:
```
// ^ Note: SharedCredentialsProvider would get renamed to SharedCredentialsCache.
// This code is using the old name to make it clearer that it already exists,
// and the rename is called out in the change checklist.
```
So the method is returning a type of a cache for which the name `create_cache` makes sense.